webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala

```diff
@@ -521,6 +416,183 @@ object ConstructResponseUtilV2 {
             case (resourceIri: IRI, resource: ResourceWithValueRdfData) if !resource.isMainResource => resourceIri
         }.toSet
 
+
+
+        // get incoming links for each resource: a map of resource IRIs to resources that link to it
+        val incomingLinksForResource: Map[IRI, RdfResources] = getIncomingLink(visibleResources, flatResourcesWithValues)
+
+
+        val mainResourcesNested: Map[IRI, ResourceWithValueRdfData] = mainResourceIrisVisible.map {
+            resourceIri =>
+                val transformedResource = nestResources(
+                    resourceIri = resourceIri,
+                    flatResourcesWithValues = flatResourcesWithValues,
+                    visibleResources = visibleResources,
+                    dependentResourceIrisVisible = dependentResourceIrisVisible,
+                    dependentResourceIrisNotVisible = dependentResourceIrisNotVisible,
+                    incomingLinksForResource = incomingLinksForResource
+                )
+
+                resourceIri -> transformedResource
+        }.toMap
+
+        MainResourcesAndValueRdfData(
+            resources = mainResourcesNested,
+            hiddenResourceIris = mainResourceIrisNotVisible ++ dependentResourceIrisNotVisible
+        )
+    }
+
+    /**
+     * This method creates a map of property IRIs to sequences of value IRIs.
+     *
+     * @param assertionsExplicit
+     * @param valueObjectIris           set of all value object IRIs
+     * @return                          the sequence of IRIs
+     */
+    private def mapIriSequence(assertionsExplicit: ConstructPredicateObjects, valueObjectIris: Set[IRI] ) :  Map[SmartIri, Seq[IRI]] =  {
+        val valuePropertyToObjectIris: Map[SmartIri, Seq[IRI]] = assertionsExplicit.map {
+            case (pred: SmartIri, objs: Seq[LiteralV2]) =>
+                // Get only the assertions in which the object is a value object IRI.
+
+                val valueObjIris: Seq[IriLiteralV2] = objs.collect {
+                    case iriObj: IriLiteralV2 if valueObjectIris(iriObj.value) => iriObj
+                }
+
+                pred -> valueObjIris
+        }.filter {
+            case (_: SmartIri, objs: Seq[IriLiteralV2]) => objs.nonEmpty
+        }.groupBy {
+            case (pred: SmartIri, _: Seq[IriLiteralV2]) =>
+                // Turn the sequence of assertions into a Map of predicate IRIs to assertions.
+                pred
+        }.map {
+            case (pred: SmartIri, valueAssertions: Map[SmartIri, Seq[IriLiteralV2]]) =>
+                // Replace the assertions with their objects, i.e. the value object IRIs.
+                pred -> valueAssertions.values.flatten.map(_.value).toSeq
+        }
+
+        return valuePropertyToObjectIris
```

Comment:
In Scala, no return statement is needed. It's enough to write `valuePropertyToObjectIris` on line 474. Or you could even omit the assignment to `valuePropertyToObjectIris` but I think it's good for clarity because you are using explicit type annotations.

Reply:
`return` is never needed in Scala, and we don't use it in Knora:

https://tpolecat.github.io/2014/05/09/return.html